# Bioinformatics_assignments
Bioinformatics assignments

Write a Python3 script named readingFrames.py that will translate a FASTA sequence of nucleotides into the amino acid sequences of all 6 possible reading frames.Your program should be able to read the sequence from a file whose filename is input as a command line argument from the command prompt:

    python readingFrames.py seq.fasta.txt

An example of the translation into the 6 reading frames is the Expasy tool (http://us.expasy.org/tools/dna.html (Links to an external site.)), which also finds open reading frames (ORFs) and color-codes them with a reddish highlight. 

Your code should output in the amino acid sequence to a plain text file named readingFrames.out.txt in the format output by Expasy  (except there would be no highlighting, no color and no bold font in the output). Ex:

5'3' Frame 1
P A R A A R G S S T S L G A S P S S A L G S T D L F S R R K V H W N W K K W I Y L P S K F K K Y K Met S F Met L C R K S Stop .....

 

An ORF is the part of a reading frame that has the potential to code for a protein or peptide. It is a continuous stretch of codons beginning with a start codon and ending with a stop codon. You are NOT required to find the ORFs, but only to provide the 6 reading frames in which an ORF could be picked out.

You may assume that the gene is in proper FASTA format (ie- it contains only letters encoding nucleotides, numbers and whitespace). You may assume the letters are restricted to capital and lowercase A,C,T and G.

Your script should be able to translate the gene found here: http://www.cs.usfca.edu/~pfrancislyon/seq.fasta.txt (Links to an external site.) 

 

Programming Style:

Encapsulate functionality, such as reading a file, as a function
Document functions (Ex: fill in the docstring auto-generated by PyCharm)
Avoid many if statements to convert 64 codons to 20 amino acids. (Ex: can use a dictionary instead)
Use the approved form for file opening:
            with open(filename, 'r') as infile:

                 seq = infile.read()

You may find the following regular expression helpful:
                   s = re.sub('[^acgtACGT]', '', seq).upper()

 

Optional:

readingFramesTemplate.pyPreview the document

re.sub(pattern, repl, string, count=0, flags=0) (Links to an external site.)

Return the string obtained by replacing the leftmost non-overlapping occurrences of pattern in string by the replacement repl

https://docs.python.org/2/library/re.html (Links to an external site.)

https://www.regular-expressions.info/quickstart.html
